File: examples/jeggs/count_graph_ops.jl
using JSON3
using Micrograd
using Statistics

include("utils.jl")
using .MicrogradExampleUtils

const DEFAULT_N_SAMPLES = 200
const DEFAULT_ARCH = [16, 16, 1]
const DEFAULT_N_STEPS = 100
const DEFAULT_ACTIVATION = :relu
const DEFAULT_LOSS = :hinge
const DEFAULT_LR = 1.0
const BENCHMARK_JSON = normpath(joinpath(@__DIR__, "..", "results", "bench_julia.json"))

function op_label(node::Value)
    name = String(nameof(typeof(node._backward)))
    if occursin("add_backward", name)
        return "+"
    elseif occursin("mul_backward", name)
        return "*"
    elseif occursin("pow_backward", name)
        return "pow"
    elseif occursin("inv_backward", name)
        return "inv"
    elseif occursin("neg_backward", name)
        return "neg"
    elseif occursin("exp_backward", name)
        return "exp"
    elseif occursin("log_backward", name)
        return "log"
    elseif occursin("tanh_backward", name)
        return "tanh"
    elseif occursin("relu_backward", name)
        return "ReLU"
    else
        return "leaf"
    end
end

function count_graph(root::Value)
    counts = Dict{String,Int}()
    for node in topo(root)
        label = op_label(node)
        counts[label] = get(counts, label, 0) + 1
    end
    return counts
end

function merge_counts!(dest::Dict{String,Int}, src::Dict{String,Int})
    for (key, value) in src
        dest[key] = get(dest, key, 0) + value
    end
    return dest
end

function benchmark_median_seconds(path::AbstractString)
    data = JSON3.read(read(path, String), Dict{String,Any})
    target_arch = full_arch(DEFAULT_ARCH)
    for result in data["results"]
        if Int(result["n_samples"]) == DEFAULT_N_SAMPLES &&
           Int(result["n_steps"]) == DEFAULT_N_STEPS &&
           Int.(result["arch"]) == target_arch &&
           String(get(result, "activation", String(DEFAULT_ACTIVATION))) == String(DEFAULT_ACTIVATION) &&
           String(get(result, "loss", String(DEFAULT_LOSS))) == String(DEFAULT_LOSS)
            return median(Float64[trial["elapsed_seconds"] for trial in result["trials"]])
        end
    end
    error("Could not find matching benchmark result in $path")
end

function main()
    X, y = load_dataset(DEFAULT_N_SAMPLES)
    model = MLP(2, DEFAULT_ARCH; activation=DEFAULT_ACTIVATION)
    load_weights!(model, weights_path(DEFAULT_ARCH))
    loss_fn = loss_function(DEFAULT_LOSS)

    total_counts = Dict{String,Int}()
    for step in 0:(DEFAULT_N_STEPS - 1)
        total_loss, _ = loss_fn(model, X, y)
        merge_counts!(total_counts, count_graph(total_loss))
        zero_grad!(model)
        backward!(total_loss)
        update_params!(model, step, DEFAULT_N_STEPS, DEFAULT_LR)
    end

    leaf_nodes = get(total_counts, "leaf", 0)
    total_nodes = sum(values(total_counts))
    op_nodes = total_nodes - leaf_nodes
    median_seconds = benchmark_median_seconds(BENCHMARK_JSON)

    summary = Dict(
        "language" => "julia",
        "config" => Dict(
            "n_samples" => DEFAULT_N_SAMPLES,
            "arch" => full_arch(DEFAULT_ARCH),
            "n_steps" => DEFAULT_N_STEPS,
            "activation" => String(DEFAULT_ACTIVATION),
            "loss" => String(DEFAULT_LOSS),
        ),
        "graph_node_counts" => Dict(total_counts),
        "total_graph_nodes" => total_nodes,
        "leaf_nodes" => leaf_nodes,
        "operation_nodes" => op_nodes,
        "benchmark_median_elapsed_seconds" => median_seconds,
        "effective_total_nodes_per_second" => total_nodes / median_seconds,
        "effective_operation_nodes_per_second" => op_nodes / median_seconds,
    )

    JSON3.pretty(stdout, summary)
    println()
end

main()

File: examples/karpathy/count_graph_ops.py
import json
from collections import Counter
from pathlib import Path
from statistics import median

from nn import MLP
from utils import full_arch, hinge_loss, load_dataset, load_weights, update_params, weights_path

DEFAULT_N_SAMPLES = 200
DEFAULT_ARCH = [16, 16, 1]
DEFAULT_N_STEPS = 100
DEFAULT_ACTIVATION = "relu"
DEFAULT_LOSS = "hinge"
DEFAULT_LR = 1.0
BENCHMARK_JSON = Path(__file__).resolve().parent.parent / "results" / "bench_python.json"


def op_label(node):
    if not node._op:
        return "leaf"
    if node._op.startswith("**"):
        return "pow"
    return node._op


def topo(root):
    order = []
    visited = set()
    stack = [(root, False)]
    while stack:
        node, expanded = stack.pop()
        if expanded:
            order.append(node)
            continue
        if node in visited:
            continue
        visited.add(node)
        stack.append((node, True))
        for child in node._prev:
            if child not in visited:
                stack.append((child, False))
    return order


def count_graph(root):
    counts = Counter()
    for node in topo(root):
        counts[op_label(node)] += 1
    return counts


def benchmark_median_seconds(path: Path) -> float:
    with path.open() as f:
        data = json.load(f)
    target_arch = full_arch(DEFAULT_ARCH)
    for result in data["results"]:
        if (
            result["n_samples"] == DEFAULT_N_SAMPLES
            and result["n_steps"] == DEFAULT_N_STEPS
            and result["arch"] == target_arch
            and result.get("activation", DEFAULT_ACTIVATION) == DEFAULT_ACTIVATION
            and result.get("loss", DEFAULT_LOSS) == DEFAULT_LOSS
        ):
            return median(trial["elapsed_seconds"] for trial in result["trials"])
    raise RuntimeError(f"Could not find matching benchmark result in {path}")


def main() -> None:
    X, y = load_dataset(DEFAULT_N_SAMPLES)
    model = MLP(2, DEFAULT_ARCH, activation=DEFAULT_ACTIVATION)
    load_weights(model, weights_path(DEFAULT_ARCH))

    total_counts = Counter()
    for step in range(DEFAULT_N_STEPS):
        total_loss, _ = hinge_loss(model, X, y)
        total_counts.update(count_graph(total_loss))
        model.zero_grad()
        total_loss.backward()
        update_params(model, step, DEFAULT_N_STEPS, DEFAULT_LR)

    leaf_nodes = total_counts.get("leaf", 0)
    total_nodes = sum(total_counts.values())
    op_nodes = total_nodes - leaf_nodes
    median_seconds = benchmark_median_seconds(BENCHMARK_JSON)

    summary = {
        "language": "python",
        "config": {
            "n_samples": DEFAULT_N_SAMPLES,
            "arch": full_arch(DEFAULT_ARCH),
            "n_steps": DEFAULT_N_STEPS,
            "activation": DEFAULT_ACTIVATION,
            "loss": DEFAULT_LOSS,
        },
        "graph_node_counts": dict(sorted(total_counts.items())),
        "total_graph_nodes": total_nodes,
        "leaf_nodes": leaf_nodes,
        "operation_nodes": op_nodes,
        "benchmark_median_elapsed_seconds": median_seconds,
        "effective_total_nodes_per_second": total_nodes / median_seconds,
        "effective_operation_nodes_per_second": op_nodes / median_seconds,
    }
    print(json.dumps(summary, indent=2))


if __name__ == "__main__":
    main()
